### DISCUSSION TITLE
legacy_model_binding and extended models dilemna

### DISCUSSION THREAD

**Discussion by uandco** [livewire/livewire#9091] [1. Help] [answered]
Hi,

I have a fairly large Livewire 2 project heavily relying on the legacy model binding. I want to convert it to Livewire 3.
`legacy_model_binding` will definitely help, but my early tests show that extended models are not compatible with this.

For example, I have a `App\ModelResources\User` extending `App\Model\User`.

In a Livewire component, I have:

```
use Illuminate\Database\Eloquent\Model;
use App\ModelResources\User;

class DynamicForm extends Component {
    public  $model;
    ...
    public function mount(int $id) {
        $this->model = User::find($id);
        ...
    }
    ...
}
```

The first page load works and displays the correct properties of the model, but any update fails because it references the base model, not the extended one.

For example calling `$model->exampleMethod()` in the associated blade template now fails because `exampleMethod` is defined on the extension, not `app\Model\User`.

It also fails to update the model properties because it's bound the the original model, not the extension.

Looking at the `livewire/update` payload, the snapshot looks like this:

```
		"model": [
			{
                             ...
			},
			{
				"key": 1,
				"class": "App\\Models\\User",
				"relations": [
                                    ...
				],
				"s": "elmdl"
			}
```

Clearly the component thinks it's dealing with `App\Models\User` instead of `App\ModelResources\User`

Turning the extension into a Trait used on the base model works, but that would involve quite a lot of rewriting.

Is this a bug or the intended behaviour with `legacy_model_binding` and extended models?

Thanks!

**Comment by joshhanley** [comment]
@uandco hmm that seems strange it's not tracking the model's correct class. I believe it should do that, so I'd suspect it's a bug. PRs welcome to fix it 🙂 You'll likely find the issue in the legacy EloquentModelSynth. If you need a work around for now, you could create a new class and extend the legacy EloquentModelSynth, register it in your `AppServiceProvider` and then override any methods causing the issue. Livewire is smart enough to use your custom registered Synth over the inbuilt one.

**Comment by uandco** [comment, marked as answer]
@joshhanley thanks for pointing me to `EloquentModelSynth`, I found the culprit:

```
public function dehydrate($target, $dehydrateChild)
    {
        $class = $target::class;

        try {
            // If no alias is found, this just returns the class name
            $alias = $target->getMorphClass();
        } catch (ClassMorphViolationException $e) {
            // If the model is not using morph classes, this exception is thrown
            $alias = $class;
        }
...
```

My extended classes do use `getMorphClass` as I need to use the base model in the morph relations, but I'm not sure why Livewire needs to redirect the target class to the morphClass one. I will attempt to override this Synthesizer and strip the `getMorphClass` logic.

**Reply by joshhanley** [reply]
@uandco we did that because people were complaining that their application structure (model classes) being sent to the front end. So instead of sending the class name, if someone has defined morph names for their classes, we use them instead. That way the payload says `post` instead of `App/Models/Post`.

**Reply by uandco** [reply]
OK, if this is all it fixes, I don't really mind bypassing it. I've just overwritten the synthetizer and it works. I guess `EloquentModelSynth` is very unlikely to change seeing this is legacy support?

**Reply by joshhanley** [reply]
@uandco yeah I've bypassed/ overridden a few of the default Synths so I can customise how they're handled in some of my apps. Yep, only bugfixes, it won't get any new changes.

**Reply by uandco** [reply]
Thanks @joshhanley now back on upgrading the app, discovering dynamic component properties are not a thing anymore, more fun in sight, ahah. Anyways, I'll close this discussion as we found where that behaviour came from. Thanks again for your help.
